### PR TITLE
Fix #623: Path rebase fails if path mentioned in SARIF file differs from resolved path only in drive letter

### LINKED ISSUE #623: Path rebase fails if path mentioned in SARIF file differs from resolved path only in drive letter
When you double-click on a result in a file that doesn't exist on your disk at the specified location, the viewer prompts with an Open File dialog. When you select the file in the correct location, the viewer should open it. But it doesn't.

Repro steps:

1. Unzip this archive to location C:\ (the root of the C: drive), which will create folder C:\EmbeddedContentDemo.

[EmbeddedContentDemo.zip](https://github.com/Microsoft/sarif-sdk/files/1322347/EmbeddedContentDemo.zip)

2. Open VS 2017 with the latest VSIX viewer installed.

3. Select menu item **Tools**, **Open Static Analysis Log File**, **Open SARIF file**. The "Open Static Analysis Results Interchange Format (SARIF) file" dialog opens.

4. Select file C:\EmbeddedContentDemo\2-IncorrectPaths.sarif, and then click **Open**. A single result appears in the Error List window.

5. Double-click the result. The "Locate missing file" dialog opens.

6. Select file C:\EmbeddedContentDemo\src\FileWithResult.cpp, and then click **Open**.

Expected: The viewer should open the selected file and highlight the line with the result.

Actual: An exception occurs:

exception

Here is the ActivityLog file mentioned in the exception (renamed to .xml.txt because you can't attach an XML file to an issue):

[ActivityLog.xml.txt](https://github.com/Microsoft/sarif-sdk/files/1322375/ActivityLog.xml.txt)

The relevant part is at the end:
```
  <entry>
    <record>650</record>
    <time>2017/09/21 18:41:51.379</time>
    <type>Error</type>
    <source>Editor or Editor Extension</source>
    <description>System.ArgumentOutOfRangeException: StartIndex cannot be less than zero.
Parameter name: startIndex
   at System.String.Substring(Int32 startIndex, Int32 length)
   at Microsoft.Sarif.Viewer.CodeAnalysisResultManager.GetRebaselinedFileName(String uriBaseId, String fileName) in E:\src\sarif-sdk\src\Sarif.Viewer.VisualStudio\CodeAnalysisResultManager.cs:line 443
   at Microsoft.Sarif.Viewer.CodeAnalysisResultManager.TryRebaselineCurrentSarifError(String uriBaseId, String originalFilename) in E:\src\sarif-sdk\src\Sarif.Viewer.VisualStudio\CodeAnalysisResultManager.cs:line 266
   at Microsoft.Sarif.Viewer.ResultTextMarker.NavigateTo(Boolean usePreviewPane) in E:\src\sarif-sdk\src\Sarif.Viewer.VisualStudio\ResultTextMarker.cs:line 71
   at Microsoft.Sarif.Viewer.ErrorList.SarifTableControlEventProcessorProvider.EventProcessor.PreprocessNavigate(ITableEntryHandle entryHandle, TableEntryNavigateEventArgs e) in E:\src\sarif-sdk\src\Sarif.Viewer.VisualStudio\ErrorList\SarifTableControlEventProcessorProvider.cs:line 94
   at Microsoft.VisualStudio.Shell.TableControl.Implementation.TableControlEventManager.&lt;&gt;c__DisplayClass40_0.&lt;OnEntryElementNavigate&gt;b__2(ITableControlEventProcessor p, ITableEntryHandle entry) 
  at Microsoft.VisualStudio.Shell.TableControl.Implementation.TableControlEventManager.&lt;&gt;c__DisplayClass60_1`1.&lt;EventProcessorHandler&gt;b__0()
   at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.CallExtensionPoint(Object errorSource, Action call)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)</description>
  </entry>

```

### PR DESCRIPTION
**Pull request by lgolding**
This is Step 3 of my plan to fix Bug #623, and get `CodeAnalysisResultManager` under test in the process. There will be four PRs, as follows:

1.	Inject dependencies into `CodeAnalysisResultManager`. **DONE** PR #670
2.	Write unit tests for `CodeAnalysisResultManager.GetRebaselinedFileName`. **DONE** PR #671
3.	Fix Bug #623 (and add unit test). **<= THIS PR**
4.	Perform further refactoring on `CodeAnalysisResultManager.GetRebaselinedFileName`.

It turns out that the bug was more general, as you will see by the last of the three unit tests I added.

### CODE REVIEW

File: src/Sarif.Viewer.VisualStudio.UnitTests/CodeAnalysisResultManagerTests.cs [hunk 1]
```diff
@@ -155,5 +155,83 @@ public void CodeAnalysisResultManager_GetRebaselinedFileName_IgnoresMismatchedFi
             Tuple<string, string>[] remappedPathPrefixes = target.GetRemappedPathPrefixes();
             remappedPathPrefixes.Should().BeEmpty();
         }
+
+        [Fact]
+        public void CodeAnalysisResultManager_GetRebaselinedFileName_WhenUserDoesNotSelectRebaselinedPath_UsesPathFromLogFile()
```
Comment (lgolding) [inline src/Sarif.Viewer.VisualStudio.UnitTests/CodeAnalysisResultManagerTests.cs]: This additional test is unrelated to the bug fix. I should have included it in the previous PR.

File: src/Sarif.Viewer.VisualStudio.UnitTests/CodeAnalysisResultManagerTests.cs [hunk 2]
```diff
@@ -155,5 +155,83 @@ public void CodeAnalysisResultManager_GetRebaselinedFileName_IgnoresMismatchedFi
...
+
+            // Assert.
+            rebaselinedFileName.Should().Be(FileNameInLogFile);
+
+            Tuple<string, string>[] remappedPathPrefixes = target.GetRemappedPathPrefixes();
+            remappedPathPrefixes.Should().BeEmpty();
+        }
+
+        [Fact]
+        public void CodeAnalysisResultManager_GetRebaselinedFileName_WhenRebaselinedPathDiffersOnlyInDriveLetter_ReturnsRebaselinedPath()
```
Comment (lgolding) [inline src/Sarif.Viewer.VisualStudio.UnitTests/CodeAnalysisResultManagerTests.cs]: This test and the next one demonstrate the fix.

File: src/Sarif.Viewer.VisualStudio.UnitTests/CodeAnalysisResultManagerTests.cs [hunk 3]
```diff
@@ -155,5 +155,83 @@ public void CodeAnalysisResultManager_GetRebaselinedFileName_IgnoresMismatchedFi
...
+            var target = new CodeAnalysisResultManager(
+                null,                               // This test never touches the file system.
+                this.keyboard,
+                this.commonDialogFactory);
+
+            // Act.
+            string rebaselinedFileName = target.GetRebaselinedFileName(uriBaseId: null, fileName: FileNameInLogFile);
+
+            // Assert.
+            rebaselinedFileName.Should().Be(RebaselinedFileName);
```
Comment (vinaykapadia) [inline src/Sarif.Viewer.VisualStudio.UnitTests/CodeAnalysisResultManagerTests.cs]: Names are really similar, sort of confusing. Not a huge issue, since it's just a small piece of code.

Comment (lgolding) [reply src/Sarif.Viewer.VisualStudio.UnitTests/CodeAnalysisResultManagerTests.cs]: @vinaykapadia  You're talking about the two variable names that differ only in case? Fair enough. I will change that variable name to `actualRebaselinedFileName` in all of the tests, and then merge. Thanks!

Comment (vinaykapadia) [reply src/Sarif.Viewer.VisualStudio.UnitTests/CodeAnalysisResultManagerTests.cs]: Yeah, they're only different by case. Again, it's not a huge deal, but when I was reading it, at first glance I thought you were comparing a string to itself.

---
In reply to: [150113292](https://github.com/Microsoft/sarif-sdk/pull/672#discussion_r150113292) [](ancestors = 150113292)
